atomate/vasp/firetasks/exchange_tasks.py

```diff
@@ -0,0 +1,458 @@
+        mmdb.collection.insert(task_doc)
+
+
+@explicit_serialize
+class HeisenbergConvergence(FiretaskBase):
+    """
+    Quick check to see if the Heisenberg model has "converged" for any particular nearest neighbor cutoff value in the sense that |J_ij| < |E_0| for all i, j. 
+
+    If not, it doesn't make sense to do Monte Carlo and the system is either 1) not well described by the Heisenberg Model (perhaps the magnetic moments are not localized) or 2) not correctly captured by the provided low-energy magnetic orderings.
+
+    Args:
+        db_file (str): path to file containing the database credentials.
+        exchange_wf_uuid (int): Unique id for record keeping.
+        parent_structure (Structure): Structure object.
+        avg (bool): <J> exchange param only.
+
+    TODO:
+        * More robust convergence check
+
+    """
+
+    required_params = ["db_file", "exchange_wf_uuid", "parent_structure", "avg"]
+    optional_params = []
+
+    def run_task(self, fw_spec):
+
+        db_file = env_chk(self["db_file"], fw_spec)
+        wf_uuid = self["exchange_wf_uuid"]
+
+        # Get Heisenberg models from db
+        mmdb = VaspCalcDb.from_db_file(db_file, admin=True)
+        mmdb.collection = mmdb.db["exchange"]
+
+        formula_pretty = self["parent_structure"].composition.reduced_formula
+
+        # Get documents
+        docs = list(
+            mmdb.collection.find(
+                {"wf_meta.wf_uuid": wf_uuid}, ["heisenberg_model", "nn_cutoff"]
+            )
+        )
+
+        hmodels = [HeisenbergModel.from_dict(d["heisenberg_model"]) for d in docs]
+        cutoffs = [d["nn_cutoff"] for d in docs]
+
+        # Take <J> only if no static calcs
+        if self["avg"]:
+            hmodel = hmodels[0].as_dict()
+        else:
+            # Check for J_ij convergence
+            converged_list = []
+            for cutoff, hmodel in zip(cutoffs, hmodels):
+                ex_params = hmodel.ex_params
+                converged = True
+
+                # J_ij exists
+                if len(ex_params) > 1:
+                    E0 = abs(ex_params["E0"])
+                    for k, v in ex_params.items():
+                        if k != "E0":
+                            if abs(v) > E0:  # |J_ij| > |E0| unphysical!
+                                converged = False
+                else:  # Only <J> was computed
+                    converged = False
+
+                if converged:
+                    converged_list.append(hmodel.as_dict())
+
+            # If multiple Heisenberg Models converged, take the maximal cutoff
+            if len(converged_list) > 0:
+                hmodel = converged_list[-1]  # Largest cutoff
+            else:  # No converged models
+                hmodel = None
+
+        # Update FW spec with converged hmodel or None
+        update_spec = {"converged_heisenberg_model": hmodel}
+
+        return FWAction(update_spec=update_spec)
+
+
+@explicit_serialize
+class VampireMC(FiretaskBase):
```

Comment:
Again, if this is likely to take some time, can we break up the running and ToDb steps.

Reply:
Yes, I'll split them up.